### ISSUE TITLE
`flutter analyze` is confused about running `pub get`

### ISSUE THREAD

**Issue by leafpetersen** [flutter/flutter#7273]
Trying out the instructions from here:

https://flutter.io/setup/

I get analysis errors until I run `flutter upgrade`, which doesn't seem to be mentioned anywhere.

Transcript as follows:

```
Last login: Mon Oct 31 16:56:31 on ttys004
leafp-macbookpro:~ leafp$ git clone https://github.com/flutter/flutter.git
Cloning into 'flutter'...
remote: Counting objects: 67773, done.
remote: Compressing objects: 100% (33/33), done.
remote: Total 67773 (delta 8), reused 0 (delta 0), pack-reused 67740
Receiving objects: 100% (67773/67773), 20.64 MiB | 1.44 MiB/s, done.
Resolving deltas: 100% (48658/48658), done.
leafp-macbookpro:~ leafp$ export PATH=`pwd`/flutter/bin:$PATH
leafp-macbookpro:~ leafp$ flutter doctor
Downloading Dart SDK 1.21.0...
######################################################################## 100.0%
Building flutter tool...
Downloading Material fonts...                       1616ms
Downloading package sky_engine...                   945ms
Running 'flutter packages get' in sky_engine...     353ms
Building Dart SDK summary...                        1510ms
Downloading engine artifacts android-arm...         7802ms
Downloading engine artifacts android-arm-profile... 2401ms
Downloading engine artifacts android-arm-release... 2302ms
Downloading engine artifacts android-x64...         3505ms
Downloading engine artifacts android-x86...         8819ms
Downloading engine artifacts ios...                 13667ms
Downloading engine artifacts ios-profile...         12624ms
Downloading engine artifacts ios-release...         9491ms
Downloading darwin-x64 tools...                     4004ms
Downloading android-arm-profile/darwin-x64 tools... 3475ms
Downloading android-arm-release/darwin-x64 tools... 1073ms
[✓] Flutter (on Mac OS, channel master)
    • Flutter at /Users/leafp/flutter
    • Framework revision 977a25f29c (37 minutes ago), 2016-12-15 13:54:53
    • Engine revision 2f32e8054d
    • Tools Dart version 1.21.0

[x] Android toolchain - develop for Android devices
    x Android Studio / Android SDK not found. Download from https://developer.android.com/sdk/
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).

[-] iOS toolchain - develop for iOS devices (Xcode 8.1)
    • XCode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.1, Build version 8B62
    x Brew not installed; use this to install tools for iOS device development.
      Download brew at http://brew.sh/.

[✓] IntelliJ IDEA Ultimate Edition (version 2016.3.1)
    • Dart plugin version 163.9166.22
    • Flutter plugin version 0.1.6

[✓] Connected devices
    • None
leafp-macbookpro:~ leafp$ flutter doctor
[✓] Flutter (on Mac OS, channel master)
    • Flutter at /Users/leafp/flutter
    • Framework revision 977a25f29c (37 minutes ago), 2016-12-15 13:54:53
    • Engine revision 2f32e8054d
    • Tools Dart version 1.21.0

[x] Android toolchain - develop for Android devices
    x Android Studio / Android SDK not found. Download from https://developer.android.com/sdk/
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).

[-] iOS toolchain - develop for iOS devices (Xcode 8.1)
    • XCode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.1, Build version 8B62
    x Brew not installed; use this to install tools for iOS device development.
      Download brew at http://brew.sh/.

[✓] IntelliJ IDEA Ultimate Edition (version 2016.3.1)
    • Dart plugin version 163.9166.22
    • Flutter plugin version 0.1.6

[✓] Connected devices
    • None
leafp-macbookpro:~ leafp$ flutter analyze flutter/packages/flutter/test/animation/animation_controller_test.dart 
Analyzing 846 files...
[warning] Undefined name 'FontWeight'. (flutter/dev/benchmarks/complex_layout/lib/main.dart, line 431, col 62)
[warning] Undefined name 'FontWeight'. (flutter/dev/benchmarks/complex_layout/lib/main.dart, line 350, col 83)
[warning] Undefined name 'FontWeight'. (flutter/dev/benchmarks/complex_layout/lib/main.dart, line 352, col 82)
[error] Undefined class 'Point'. (flutter/dev/benchmarks/microbenchmarks/lib/gestures/data/velocity_tracker_data.dart, line 927, col 21)

/// Many more errors elided

leafp-macbookpro:~ leafp$ flutter upgrade
Upgrading Flutter from /Users/leafp/flutter...
Already up-to-date.
Running 'flutter packages get' in flutter...        5298ms
Building unlinked bundles for flutter...            1966ms
Running 'flutter packages get' in flutter_driver... 5112ms
Building unlinked bundles for flutter_driver...      84ms
Running 'flutter packages get' in flutter_markdown... 4717ms
Building unlinked bundles for flutter_markdown...    48ms
Running 'flutter packages get' in flutter_test...   4477ms
Building unlinked bundles for flutter_test...        89ms
Building unlinked bundles for flutter_tools...      305ms
Building unlinked bundles for sky_engine...          25ms

Upgrading engine...
Already up-to-date.

Flutter • channel master • https://github.com/flutter/flutter.git
Framework • revision 977a25f29c (38 minutes ago) • 2016-12-15 13:54:53
Engine • revision 2f32e8054d
Tools • Dart 1.21.0

Running flutter doctor...
[✓] Flutter (on Mac OS, channel master)
    • Flutter at /Users/leafp/flutter
    • Framework revision 977a25f29c (38 minutes ago), 2016-12-15 13:54:53
    • Engine revision 2f32e8054d
    • Tools Dart version 1.21.0

[x] Android toolchain - develop for Android devices
    x Android Studio / Android SDK not found. Download from https://developer.android.com/sdk/
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).

[-] iOS toolchain - develop for iOS devices (Xcode 8.1)
    • XCode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.1, Build version 8B62
    x Brew not installed; use this to install tools for iOS device development.
      Download brew at http://brew.sh/.

[✓] IntelliJ IDEA Ultimate Edition (version 2016.3.1)
    • Dart plugin version 163.9166.22
    • Flutter plugin version 0.1.6

[✓] Connected devices
    • None
leafp-macbookpro:~ leafp$ flutter analyze flutter/packages/flutter/test/animation/animation_controller_test.dart 
Analyzing 846 files...
No analyzer warnings! (ran in 23.6s; 56 public members lack documentation)
leafp-macbookpro:~ leafp$ 

```

**Comment by Hixie** [comment]
The problem is that `pub get` doesn't get run in the flutter repo by default (since for most people it's a waste of time).

What I don't understand is what the flutter tool thinks it's analyzing in your case. If you'd run it from inside the package you wanted to analyze a file from, it would have run pub get for you and then only analyzed that package. But because you weren't in a package, it just crawled the whole directory and analyzed every dart file it found (other than those in directories with `.dartignore` files). I think it's creating a temporary file and a temporary `.packages` from the `.packages` files it finds, but it's not clear to me why it wouldn't then run `pub get` in those directories first.

Another bug here is that the argument is ignored. You can't use `flutter analyze` to analyze a single file. We should show a message to that effect (or support analyzing a single file or path).

Another bug is that `flutter upgrade` for some reason _does_ run `pub get` in the `flutter` directory. It shouldn't do that (or we should always do it, but that seems wrong).

So in summary:
- [ ] `flutter analyze` should either complain about being given a file argument or do something with it.
- [ ] `flutter analyze` should run `pub get` in the directories it scans that contain `pubspec.yaml` files.
- [ ] `flutter upgrade` should not run `pub get` in flutter repo directories.

**Comment by Hixie** [comment]
The instructions on `flutter.io` are aimed at people writing flutter apps, not people writing flutter. For the latter, please see https://github.com/flutter/flutter/blob/master/CONTRIBUTING.md.

I've filed separate bugs for the issues I listed above, so I'm going to close this bug. Please don't hesitate to reopen if you think I've missed a key issue here.

**Comment by leafpetersen** [comment]
No problem, thanks for looking into it.
